Indexing/TermSubstitutionTree.cpp

```diff
@@ -35,428 +24,48 @@ using namespace Lib;
-      normTerm = t.term();
-    }
-
-    BindingMap svBindings;
-    getBindings(normTerm, svBindings);
-
-    unsigned rootNodeIndex=getRootNodeIndex(normTerm);
-
-    if(insert) {
-      SubstitutionTree::insert(&_nodes[rootNodeIndex], svBindings, ld);
-    } else {
-      SubstitutionTree::remove(&_nodes[rootNodeIndex], svBindings, ld);
-    }
-  }
-}
-
-TermQueryResultIterator TermSubstitutionTree::getUnifications(TermList t,
-	  bool retrieveSubstitutions)
-{
-  CALL("TermSubstitutionTree::getUnifications");
-  if(t.isOrdinaryVar()) {
-    return getAllUnifyingIterator(t,retrieveSubstitutions,false);
-  } else {
-    ASS(t.isTerm());
-    if(_vars.isEmpty()) {
-      // false here means without constraints
-      return getResultIterator<UnificationsIterator>(t.term(), retrieveSubstitutions,false);
-    } else {
-      return pvi( getConcatenatedIterator(
-          // false here means without constraints
-	  ldIteratorToTQRIterator(LDSkipList::RefIterator(_vars), t, retrieveSubstitutions,false),
-          // false here means without constraints
-	  getResultIterator<UnificationsIterator>(t.term(), retrieveSubstitutions,false)) );
-    }
-  }
-}
-
-//higher-order concern
-TermQueryResultIterator TermSubstitutionTree::getUnificationsUsingSorts(TermList t, TermList sort,
-    bool retrieveSubstitutions)
-{
-  CALL("TermSubstitutionTree::getUnificationsUsingSorts");
-
-  ASS(_extByAbs);
-
-  bool sortVar = sort.isVar();
-  bool sortArrow = sort.isArrowSort();
-  bool sortAtomic = !sortVar && !sortArrow;
-
-  if(t.isOrdinaryVar()) {
-    auto it1 = sortVar || sortArrow ? _funcSubtermsByType->getUnifications(sort, t, retrieveSubstitutions):
-               TermQueryResultIterator::getEmpty();
-    auto it2 = sortVar || sortAtomic ? getAllUnifyingIterator(t,retrieveSubstitutions,false):
-               TermQueryResultIterator::getEmpty();
-    return pvi(getConcatenatedIterator(it1, it2));
-    //TODO vars?
-  } else {
-    ASS(t.isTerm());
-    //TODO Is it OK to use t below?
-    auto it1 = _vars.isEmpty() ? TermQueryResultIterator::getEmpty() :
-               ldIteratorToTQRIterator(LDSkipList::RefIterator(_vars), t, retrieveSubstitutions,false);
-
-    auto it2 = sortVar || sortArrow ? _funcSubtermsByType->getUnifications(sort, t, retrieveSubstitutions) :
-               TermQueryResultIterator::getEmpty();
-    auto it3 = sortVar || sortAtomic ? getResultIterator<UnificationsIterator>(t.term(), retrieveSubstitutions,false):
-               TermQueryResultIterator::getEmpty();
-    return pvi(getConcatenatedIterator(getConcatenatedIterator(it1, it2), it3));
-  }
-}
-
-//TODO code sharing with getUnifications
-TermQueryResultIterator TermSubstitutionTree::getUnificationsWithConstraints(TermList t,
-          bool retrieveSubstitutions)
-{
-  CALL("TermSubstitutionTree::getUnificationsWithConstraints");
+template<class TypedOrUntypedTermList>
```

Comment:
I can't figure out how this links, since we don't provide manual template specializations. But any hypothetical linkage problems would be fixed by lifting it to the header file.